### ISSUE TITLE
Too many request-ids

### ISSUE THREAD

**Issue by rwz** [interagent/pliny#212]
Right now platform apps are getting request-id header automatically coming from hermes. Then Pliny adds it's own request id in the request-id middleware. That means that the minimum amount of request ids per single request for a platform app is now two.

What do you think about making pliny add a new request id only if there're no request ids already? That will make it reuse hermes request id when running on the platform and actually generate one when running standalone.

I can send a PR.

**Comment by tmaier** [comment]
I don't know Hermes, but this sounds reasonable.

Shall the request I'd middleware be configurable which header to use or shall it even default to the header of Hermes?

**Comment by ys** [comment]
Maybe this could be a flag, `use_exertnal_if_exists`.

Because there is internal request-ids and external ones. You might want internal to use your own even if somebody passes you one.

**Comment by mikehale** [comment]
> Because there is internal request-ids and external ones. You might want internal to use your own even if somebody passes you one.

Right. I don't know that it is possible to distinguish between a request-id set by hermes and a request-id or list of request-ids that came from farther up stream. IIRC hermes only sets a request-id if one was not provided in the original request.

More generally it is nice to have original and per-service request ids. That makes it possible to see the entire scope of a request through all services, and the lifecycle of a request in a specific service.

See https://brandur.org/request-ids#composition

**Comment by pedro** [comment]
I'm not super familiar with the motivation for both ids. Is it different than just looking for the global id and scoping by the component you're trying to search? Are there any other factors at play here, like not trusting clients to generate unique ids? /cc @brandur

**Comment by brandur** [comment]
Right, like @mikehale said, the motivation was so that in system with a lot of re-entrant calls (i.e. Heroku), it's always possible to isolate any particular subtree of requests because each request generates a new identifier and distributes it to its children along with any identifiers that it received.

```
                                         ┌─────────────────┐ 
            ┌──────────────┐             │                 │░
            │              │             │ Request ID Tree │░
            │     req1     │             │                 │░
            │              │             └─────────────────┘░
            └──────────────┘              ░░░░░░░░░░░░░░░░░░░
                    │                                        
        ┌───────────┴────────────┐                           
        ▼                        ▼                           
┌──────────────┐         ┌──────────────┐                    
│              │         │              │                    
│  req1,req2   │         │  req1,req3   │                    
│              │         │              │                    
└──────────────┘         └──────────────┘                    
                                 │                           
                    ┌────────────┴───────────┐               
                    ▼                        ▼               
            ┌──────────────┐         ┌──────────────┐        
            │              │         │              │        
            │req1,req3,req4│         │req1,req3,req5│        
            │              │         │              │        
            └──────────────┘         └──────────────┘        
```

Unfortunately, Hermes makes this problematic because you can't tell if an incoming request was meant to come to you or whether it's just Hermes sending you garbage.

Two possible ways forward:

1. I lied a little above. You can tell a request that Hermes sends you because it `X-` prefixes its request IDs. Hopefully most other things aren't doing that.
2. An option flag like @ys suggested.

**Comment by gudmundur** [comment]
@brandur I'm assuming that if you provide a `X-Request-Id` header, Hermes won't assign one.

**Comment by gudmundur** [comment]
I think this also boils down to trust, can Pliny trust the `Request-Id` that was passed to it? A malicious user, can always provide the same `Request-Id`, making it hard to parse what happened in a single request. Without adding mechanisms to sign the header, and verifying the signature, I think leaving this as is makes sense. What do you think @rwz?

**Comment by rwz** [comment]
@gudmundur yeah, I guess. I never actually thought about this from the "malicious user" perspective. Makes sense.
